### ISSUE TITLE
Neural network for fit_textured_mesh.ipynb always predicts a sphere

### ISSUE THREAD

**Issue by stephentandjiria** [facebookresearch/pytorch3d#680]
## 🐛 Bugs / Unexpected behaviors
Hi, I have been trying to implement a neural network for fit_textured_mesh.ipynb using the architecture in the Mesh-RCNN, and the resulting vertices are used to offset an ico_sphere to become a mesh prediction. However, this prediction always becomes a sphere:
![image](https://user-images.githubusercontent.com/75295743/118629157-8587b280-b7f7-11eb-87dc-b20ea6ae374f.png)
![image](https://user-images.githubusercontent.com/75295743/118629195-8ddfed80-b7f7-11eb-9320-3a1429b029da.png)
![image](https://user-images.githubusercontent.com/75295743/118629300-aa7c2580-b7f7-11eb-916f-32ee27f40e85.png)
![image](https://user-images.githubusercontent.com/75295743/118629323-ae0fac80-b7f7-11eb-8fb9-9d8ab7502fd1.png)
![image](https://user-images.githubusercontent.com/75295743/118629341-b23bca00-b7f7-11eb-867a-c7cc23a0ac03.png)

(and does not change anymore even after 10000 iterations)

## Instructions To Reproduce the Issue:

Please include the following (depending on what the issue is):

1. Any changes you made (`git diff`) or code you wrote
Complete repo: https://github.com/stephentandjiria/SphereGCN_test
Summary of [SphereGCN v2.ipynb](https://github.com/stephentandjiria/SphereGCN_test/blob/main/SphereGCN%20v2.ipynb):
```
# We initialize the source shape to be a sphere of radius 1.  
src_mesh = ico_sphere(4, device)

# ------ MODEL ------
# (2 - 5) RESNET50
resnet50_head = Resnet50Head()

if torch.cuda.is_available():
    input_batch = input_batch.to('cuda')
    resnet50_head.to('cuda')

# (6 - 7) BILINEAR INTERPOLATION - VOXEL PREDICTION
voxel_head = VoxelHead()

if torch.cuda.is_available():
    voxel_head.to('cuda')

# (8) CUBIFY
def _add_dummies(meshes):
    N = len(meshes)
    def _dummy_mesh(N, device):
        verts_batch = torch.randn(N, 4, 3, device=device)
        faces = [[0, 1, 2], [0, 2, 3], [0, 3, 1], [1, 3, 2]]
        faces = torch.tensor(faces, dtype=torch.int64)
        faces_batch = faces.view(1, 4, 3).expand(N, 4, 3).to(device)
        return Meshes(verts=verts_batch, faces=faces_batch)
    dummies = _dummy_mesh(N, meshes.device)
    verts_list = meshes.verts_list()
    faces_list = meshes.faces_list()
    for i in range(N):
        if faces_list[i].shape[0] == 0:
            # print('Adding dummmy mesh at index ', i)
            vv, ff = dummies.get_mesh(i)
            verts_list[i] = vv
            faces_list[i] = ff
    return Meshes(verts=verts_list, faces=faces_list)

def cubify_custom(voxel_scores):
    V = 48 # voxel_size
    N = voxel_scores.shape[0]
    cubify_threshold = 0.80
    voxel_probs = voxel_scores.sigmoid()
    active_voxels = voxel_probs > cubify_threshold
    voxels_per_mesh = (active_voxels.view(N, -1).sum(dim=1)).tolist()
    start = V // 4
    stop = start + V // 2
    for i in range(N):
        if voxels_per_mesh[i] == 0:
            voxel_probs[i, start:stop, start:stop, start:stop] = 1
    meshes = cubify(voxel_probs, cubify_threshold)
    meshes = _add_dummies(meshes)
    meshes = voxel_to_world(meshes)
    return meshes

# (9 - 11) REFINEMENT STAGES 1 - 3
mesh_head = MeshRefinementHead()
if torch.cuda.is_available():
    mesh_head.to('cuda')

N = input_batch.shape[0]
# cubify_custom.register_buffer("K", get_blender_intrinsic_matrix())
K = get_blender_intrinsic_matrix()

def _get_projection_matrix(N, device):
    return K[None].repeat(N, 1, 1).to(device).detach()

P = _get_projection_matrix(N, device)
```
Training Loop
```
loop = tqdm(range(Niter))
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

for i in loop:
    # Initialize optimizer
    optimizer.zero_grad()
    
    # ------ MODEL ------
    # (2 - 5) RESNET50
    img_feats = resnet50_head(input_batch)

    # (6 - 7) BILINEAR INTERPOLATION - VOXEL PREDICTION    
    voxel_scores = voxel_head(img_feats[-1])

    # (8) CUBIFY
    meshes_out = cubify_custom(voxel_scores)
    
    # (9 - 11) REFINEMENT STAGES 1 - 3
    refined_meshes = mesh_head(img_feats, src_mesh, P)
    meshes_out = refined_meshes[-1]

    # DEFORM MESH
    new_src_mesh = src_mesh.offset_verts(meshes_out.verts_packed())
    #new_src_mesh = src_mesh.offset_verts(deform_verts)
    new_src_mesh.textures = meshes_out.textures
    
    # Losses to smooth /regularize the mesh shape
    loss = {k: torch.tensor(0.0, device=device) for k in losses}
    update_mesh_shape_prior_losses(new_src_mesh, loss)
    
    # Randomly select two views to optimize over in this iteration.  Compared
    # to using just one view, this helps resolve ambiguities between updating
    # mesh shape vs. updating mesh texture
    for j in np.random.permutation(num_views).tolist()[:num_views_per_iteration]:
        images_predicted = renderer_textured(new_src_mesh, cameras=target_cameras[j], lights=lights)

        # Squared L2 distance between the predicted silhouette and the target 
        # silhouette from our dataset
        predicted_silhouette = images_predicted[..., 3]
        loss_silhouette = ((predicted_silhouette - target_silhouette[j]) ** 2).mean()
        loss["silhouette"] += loss_silhouette / num_views_per_iteration

        # Squared L2 distance between the predicted RGB image and the target 
        # image from our dataset
        predicted_rgb = images_predicted[..., :3]
        loss_rgb = ((predicted_rgb - target_rgb[j]) ** 2).mean()
        loss["rgb"] += loss_rgb / num_views_per_iteration
    
    # Weighted sum of the losses
    sum_loss = torch.tensor(0.0, device=device)
    for k, l in loss.items():
        sum_loss += l * losses[k]["weight"]
        losses[k]["values"].append(l)
    
    # print(new_src_mesh.verts_packed())
    
    # Print the losses
    loop.set_description("total_loss = %.6f" % loss["edge"])
    #loop.set_description("total_loss = %.6f" % sum_loss)
    
    # Plot mesh
    if i % plot_period == 0:
        visualize_prediction(new_src_mesh, renderer=renderer_textured, title="iter: %d" % i, silhouette=False)
    
    # Optimization step
    sum_loss.backward()
    optimizer.step()
```
where voxel_head.py, utils_coords.py, and mesh_head.py were from https://github.com/facebookresearch/meshrcnn

except that in the mesh_head.py, I added [texture from sigmoid output of vert_feats](https://github.com/stephentandjiria/SphereGCN_test/blob/main/mesh_head.py#L133) as suggested in the [paper](https://arxiv.org/pdf/2007.08501.pdf):

2. What you observed (including the full logs):
No error occurred, but the predicted meshes always converges to a sphere as shown above, sometimes located on the top-right, sometimes centred/left/etc, and its color is also quite random, often white, but sometimes also red/purple/etc. So, I tried to see each model's parameters before and after the training loop to see whether the training changes the parameters. Strangely only mesh_head.parameters() change:
```
#before training loop
for param in mesh_head.parameters():
    print(param)

#returns:
Parameter containing:
tensor([[ 0.0058,  0.0128,  0.0072,  ...,  0.0009, -0.0113,  0.0080],
        [ 0.0084,  0.0021, -0.0127,  ...,  0.0087, -0.0006,  0.0047],
        [-0.0079,  0.0111,  0.0159,  ..., -0.0129,  0.0058,  0.0034],
        ...,
        [-0.0054,  0.0070, -0.0092,  ...,  0.0046,  0.0068,  0.0143],
        [-0.0072, -0.0036,  0.0139,  ..., -0.0007,  0.0188,  0.0036],
        [ 0.0163, -0.0002,  0.0012,  ..., -0.0011, -0.0063,  0.0044]],
       device='cuda:0', requires_grad=True)
Parameter containing:
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.], device='cuda:0', requires_grad=True)
...
```
whereas **after** training loop:
```
for param in mesh_head.parameters():
    print(param)

#returns:
Parameter containing:
tensor([[ 0.0099,  0.0191,  0.0097,  ...,  0.0056,  0.0083,  0.0154],
        [ 0.0175,  0.0201, -0.0031,  ...,  0.0259,  0.0490,  0.0472],
        [-0.0064,  0.0151,  0.0175,  ..., -0.0088,  0.0171,  0.0076],
        ...,
        [-0.0064,  0.0044, -0.0097,  ...,  0.0014, -0.0017,  0.0138],
        [-0.0072, -0.0036,  0.0139,  ..., -0.0004,  0.0190,  0.0039],
        [ 0.0129, -0.0048, -0.0007,  ..., -0.0023, -0.0183, -0.0030]],
       device='cuda:0', requires_grad=True)
Parameter containing:
tensor([-7.0389e-02,  1.5454e-01,  7.5771e-03, -8.7144e-02,  1.1653e-01,
        -6.2932e-03,  1.5008e-02,  6.9484e-02, -1.3141e-02,  3.6608e-03,
        -2.8928e-02,  7.8971e-02, -2.0037e-03,  1.0394e-02,  0.0000e+00,
        -1.5591e-02, -3.4918e-03, -6.9734e-02, -4.1455e-02, -6.6652e-02,
         3.0146e-05, -1.7176e-02,  7.2036e-04,  0.0000e+00,  3.6719e-05,
        -2.5239e-02,  5.2621e-02,  2.2906e-02, -6.3607e-03,  8.9231e-03,
         3.2988e-02,  2.8076e-03,  5.1567e-05,  5.5176e-04,  2.0865e-02,
         1.6713e-02, -3.2546e-02, -1.5848e-04, -3.3205e-02, -1.9055e-03,
        -1.8998e-02, -7.7304e-02,  1.1051e-01,  4.8468e-02, -1.2504e-03,
         7.0037e-03, -3.4719e-02,  7.9882e-02,  3.8478e-02, -9.1704e-02,
        -5.0891e-02, -1.8442e-02,  7.0101e-02,  2.0684e-02,  3.6014e-02,
         5.9899e-03, -8.3786e-03, -1.9088e-02,  5.6254e-03, -6.5850e-02,
         1.9849e-02,  4.6223e-02, -2.2921e-02, -7.3261e-02, -8.3129e-03,
         2.9369e-02,  2.5939e-02, -1.9806e-02,  1.4967e-02,  7.1297e-02,
         0.0000e+00,  2.9648e-02, -5.6977e-02,  9.8663e-05,  4.4010e-04,
         9.1216e-03,  1.4399e-01, -5.8336e-02, -6.4269e-03, -3.1328e-04,
        -5.8733e-03,  8.6436e-02, -3.0806e-02,  5.7678e-02, -1.9527e-02,
         2.5210e-02, -9.0638e-02, -5.7443e-02,  1.7901e-02, -3.7730e-03,
         5.4724e-02, -5.4234e-02, -3.6327e-05, -5.1406e-02, -1.0442e-02,
         2.2332e-02,  3.8113e-02,  1.8557e-02, -1.9700e-02, -1.5238e-01,
         2.6085e-02, -3.6432e-02, -1.4966e-02,  6.4356e-02, -3.9371e-04,
        -3.8638e-02,  6.3183e-04, -2.1640e-02, -3.7383e-02,  0.0000e+00,
        -8.1570e-02, -3.6421e-02,  8.0950e-03,  3.9982e-02, -2.5532e-02,
        -9.7813e-02, -2.8530e-02,  2.0315e-02,  1.3823e-01,  2.1046e-02,
         7.8227e-04, -5.3432e-02,  9.5142e-03,  2.9942e-02, -6.1896e-02,
        -1.0500e-02,  1.2397e-03, -1.1859e-02], device='cuda:0',
       requires_grad=True)
...
```
as you can see the mesh_head parameters value changed, but this is not the case with resnet50_head and voxel_head, their parameters values do not change. (full output in the notebook: Section 4H is **before** training loop, Section 6 is **after** training loop)

I also checked requires_grad, but all params return True:
```
for param in resnet50_head.parameters():
    print(param.requires_grad)

for param in voxel_head.parameters():
    print(param.requires_grad)

for param in mesh_head.parameters():
    print(param.requires_grad)
```

After trying to debug for a week, I am really desperate. So, I would really appreciate if anyone could help!
Thank you!

**Comment by sainatarajan** [comment]
@stephentandjiria @nikhilaravi  Any updates regarding this?
